### DISCUSSION TITLE
JS and Toga's webview call each other

### DISCUSSION THREAD

**Discussion by benmaxyoung** [beeware/toga#2278] [Debugging help]
Function I want to achieve:
1. Use toga as a cross-platform web shell (platforms include windows, macOS, ios, android)
2. Use the front-end build package as a local resource and use it in the shell.

doubt:
1. Toga's webview does not support js calling python. Is there a solution that can support python and js to communicate with each other?
2. We have a requirement for sliding gestures on the sidebar of the screen. Does Toga support it?
3. If you use toga's webview, after the webview loads the local html resource, the post request issued will lose the postbody, and the cookie cannot be used normally. How to solve it?

**Comment by benmaxyoung** [comment]
I tried to use pywebview, but the ios and android platforms are not supported for the time being. And cannot be used with toga.
I have development experience in ios and android, but I don't know much about the principles of toga. I don’t know if I can develop toga’s webview so that JS and python can communicate with each other.

**Comment by benmaxyoung** [comment]
I think briefcase and toga are the most useful tools, so I don’t want to give them up.

**Comment by freakboy3742** [comment]
> 1. Toga's webview does not support js calling python. Is there a solution that can support python and js to communicate with each other?

See #2268.

> 2. We have a requirement for sliding gestures on the sidebar of the screen. Does Toga support it?

I have no idea what you're referring here. Are you referring to the webview not responding to gestures? Or something else?

> 3. If you use toga's webview, after the webview loads the local html resource, the post request issued will lose the postbody, and the cookie cannot be used normally. How to solve it?

Again, I have no idea what you're referring to here. The webview behaves like a webview. Cookies should be preserved in whatever context a webview will retain cookies. If you believe you've found a bug with cookie handling, we'll need a specific reproduction case for that problem.

**Comment by benmaxyoung** [comment]
I don't understand that https://github.com/beeware/toga/issues/2268. Do you mean that js calling python is in the plan?

**Reply by freakboy3742** [reply]
It has been requested as a feature, and we've agreed that it needs to be added as a feature. Now we need someone to implement the feature. I can't say it's high on my personal priority list, though.

**Reply by benmaxyoung** [reply]
Okay, because I'm not very familiar with windows, otherwise I'd love to join the program

**Comment by bennybi** [comment]
Hello, i think open a websocket service to communicate with the web app could be the solution.

**Reply by freakboy3742** [reply]
Only in the sense of a "nasty hack that you could probably make work". As [I've indicated on the linked ticket](https://github.com/beeware/toga/issues/2268#issuecomment-1849112672), there's native APIs to achieve this - we need to provide a Toga API for accessing those APIs.

**Comment by mymane1433223** [comment]
flask
